core/src/main/java/org/frankframework/http/authentication/AbstractOauthAuthenticator.java

```diff
@@ -0,0 +1,140 @@
+import static org.frankframework.util.StreamUtil.DEFAULT_INPUT_STREAM_ENCODING;
+
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.http.NameValuePair;
+import org.apache.http.auth.Credentials;
+import org.apache.http.client.entity.UrlEncodedFormEntity;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpEntityEnclosingRequestBase;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.client.methods.HttpRequestBase;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.util.EntityUtils;
+
+import lombok.extern.log4j.Log4j2;
+
+import org.frankframework.http.AbstractHttpSession;
+import org.frankframework.task.TimeoutGuard;
+import org.frankframework.util.DateFormatUtils;
+import org.frankframework.util.JacksonUtils;
+
+@Log4j2
+public abstract class AbstractOauthAuthenticator implements IOauthAuthenticator {
+
+	protected final AbstractHttpSession session;
+	protected final URI authorizationEndpoint;
+	protected final int overwriteExpiryMs;
+
+	private String accessToken;
+	private long accessTokenRefreshTime;
+
+	AbstractOauthAuthenticator(final AbstractHttpSession session) throws HttpAuthenticationException {
+		this.session = session;
+		try {
+			this.authorizationEndpoint = new URI(session.getTokenEndpoint());
+		} catch (URISyntaxException e) {
+			throw new HttpAuthenticationException(e);
+		}
+		this.overwriteExpiryMs = session.getTokenExpiry() * 1000;
+	}
+
+	protected abstract HttpEntityEnclosingRequestBase createRequest(Credentials credentials, List<NameValuePair> parameters) throws HttpAuthenticationException;
+
+	protected HttpEntityEnclosingRequestBase createPostRequestWithForm(URI uri, List<NameValuePair> formParameters) throws HttpAuthenticationException {
+		try {
+			UrlEncodedFormEntity body = new UrlEncodedFormEntity(formParameters, DEFAULT_INPUT_STREAM_ENCODING);
+
+			HttpPost request = new HttpPost(uri);
+			request.addHeader(body.getContentType());
+			request.setEntity(body);
+
+			return request;
+		} catch (UnsupportedEncodingException e) {
+			throw new HttpAuthenticationException(e);
+		}
+	}
+
+	private void refreshAccessToken(Credentials credentials) throws HttpAuthenticationException {
+		log.debug("Refreshing access token");
+		HttpRequestBase request = createRequest(credentials, new ArrayList<>());
+
+		CloseableHttpClient apacheHttpClient = session.getHttpClient();
+		TimeoutGuard tg = new TimeoutGuard(1 + session.getTimeout() / 1000, "token retrieval") {
+
+			@Override
+			protected void abort() {
+				request.abort();
+			}
+
+		};
+
+		try (CloseableHttpResponse response = apacheHttpClient.execute(request)) {
+			String responseBody = EntityUtils.toString(response.getEntity());
+
+			if (response.getStatusLine().getStatusCode() != 200) {
+				tg.cancel();
```

Comment:
Dit was de missende `tg.cancel()`.
Ik vond het niet de moeite waard om hem te checken en een timeout-exception te gooien, en daarmee dan de status-code line in de exception kwijt te raken.

De `tg.cancel()` miste omdat ik bij Thomas erop aangedrongen had, het niet in de `finally` te doen. 🫣